### ISSUE TITLE
Add a protected route

### ISSUE THREAD

**Issue by RicCu** [CaptainCodeman/sveltekit-example#1]
Thanks so much for this example!

I'm starting to play around with Svelte and your setup with firebase auth in a store is very helpful. However I haven't been able to set up a properly protected route. I tried setting it up in the `load` function like in the real-world example, but stores aren't available in `<script context="module">`s. Then I tried setting up a redirect with `goto` inside `onMount` but the store is always unset at the beginning, so it always ends up trying to redirect as if there was no user signed in. It isn't until after the `onMount` runs that the `onAuthStateChanged` in `listen` kicks in and sets the auth store to the current user.
I'm wondering if I need to store the auth value in a cookie on top of the store or if there's a better, svelte-way of doing things.
Could you add an example of a protected route in this example app?

**Comment by CaptainCodeman** [comment]
Because of how Firebase works and also because the scripts themselves are being lazy-loaded, the auth resolution is happening independently of any component `onMount` - when the component is mounted the auth status _may_ or _may not_ be known.

As it isn't guaranteed to run before or after it's better to think in terms of subscribing to the auth store and taking appropriate action when the status is known. You also need to decide what should happen _before_ the status is known.

Exactly what that action should be will depend a lot on your specific app - do you want the user to be able to navigate to a route and be told they don't have access, or be redirected someplace else and so on. It will also depend how that is decided - a simple case would be just whether they are signed in or not, a more complex example would look at their auth claims and whether they have access to that specific thing. That then gets into other similar requirements, such as "hiding the edit button if they don't have an editor role".

For the simple cases, I would think about putting the auth "guard" in the `$layout.svelte` file which can render the `<slot>` to show the current route or not, perhaps showing an 'access denied' message or redirecting if they don't.

For adapting the UI based on claims / roles I've usually used a wrapper element or an action that can be applied to individual elements.

I'll try to add some examples of both.

**Comment by quantuminformation** [comment]
Looking fwd to the auth guard.

**Comment by CaptainCodeman** [comment]
The `Auth.svelte` control shows how you can use the auth state - if you wanted to prevent something from displaying for instance, you could put something like this in a __layout and only output the slot if they are signed in. You can also check for the user having specific auth claims to provide role based access to things.

**Comment by ivanddm** [comment]
Hi,

I was trying to do this actually and finally found an approach here: [Why is my currentUser == null in Firebase Auth? | by Doug Stevenson | Firebase Developers | Medium](https://medium.com/firebase-developers/why-is-my-currentuser-null-in-firebase-auth-4701791f74f0), which is actually mentioned in the firebase docs [Manage Users in Firebase](https://firebase.google.com/docs/auth/web/manage-users)

Basically, you need to add a`onAuthStateChanged` observer to your route or `__layout`, e.g. wrapped in `onMount`, and then if `user`, redirect to another route using `goto`, something like this:

```javascript
import { app } from '$lib/firebase'

onMount(() => {
  onAuthStateChanged(auth.userAuth, async (user) => {
    if (!user) {
      await goto('/login');
    }
  });
});
```

**Comment by CaptainCodeman** [comment]
The auth store is already checking for auth state changes and providing the auth status - so there is no need to duplicate this, simply check the auth store state. If you do it in a __layout you can either allow the <slot>'d content or display a login form, or redirect to one.

**Comment by ivanddm** [comment]
do you mean, checking for auth state with `$auth.user` or `$auth.known`? I tried that and the status is `null` after refreshing the page for a split of a second (what they call in the docs an intermediate state) before it gives the actual status, and so the route ends up being always redirected.

**Comment by CaptainCodeman** [comment]
Yes, the `$auth.known` tells you whether the state of `$auth.user` has been determined yet, which isn't instant (because the client SDK has to initialize and check it's state), so either combine or nest them (the not-know state could be a loading spinner for instance).

**Comment by ivanddm** [comment]
yes, in the html components, `$auth.know` and `$auth.user` accomplishes what you say but only to hide or display content, but what I need is the route to be redirected to another route/page if the user is not signed in. For that, apparently, `onAuthStateChanged` call is needed on any page on which you want to avoid the intermediate user state, which is unfortunate. This is from the documentation: "[For each of your app's pages that need information about the signed-in user, attach an observer to the global authentication object](https://firebase.google.com/docs/auth/web/start)..."

**Comment by CaptainCodeman** [comment]
You're not understanding what I'm saying - that doc is telling you how to do it with _just_ the firebase SDK in a generic way, the auth code in this project is wrapping that and providing it in a svelte-subscribeable store. No need to repeat it, it's already provided in the store and means you only need a single place to check it.

https://github.com/CaptainCodeman/sveltekit-example/blob/master/src/lib/auth.ts#L17

**Comment by ivanddm** [comment]
I understand what you mean and I know that in `auth.ts` there is a call to `onAuthStateChange`, but that call is not enough for checking the state of the user and redirecting a route. For example, you can try that the following route guard will not work, say in a new route  `routes/about.svelte`:

```javascript
<script>
$: if (!$auth.user) {
  goto('/')
}
</ script>

<div>
...
</div>
```

when you refresh the page `/about`, the route will always redirect to `/`, even if the user is logged in, which of course is not what we want if the user is logged in. And this is due to `$auth.user` being `null` for a fraction of a second until the connection to firebase is reestablished and the user state is verified.

Anyways, you can check that if you want, and perhaps there are other ways of getting around that issue but in my understanding, you need another call to `onAuthStateChange` to omit the intermediate `null` state of the user.

**Comment by CaptainCodeman** [comment]
Well, you're not checking for if the auth status is `known`. If you follow what I've already said, several times, you'll see it works fine.

As already stated, you _don't_ need to add any additional `onAuthStateChange` subscriptions. The auth state is already encapsulated in and available via the auth store to svelte routes and components.

If you want to ignore part of the state and claim it doesn't work, that's up to you. I have better things todo than keep tryting to explain the same thing to you.

**Comment by ivanddm** [comment]
Well, I was not looking for an explanation from your side, was I? if you read my comments carefully you would realize that. If you disagree with my comments regarding the issue that @RicCu raised in the first place is completely app to you, and yea, I can say the same thing, I also have betters things to do than trying to clarify to you that your answer does not solve this issue. But anyway, we all have the right to disagree, otherwise, I would suggest that you don't post public repos or at least don't allow to open issues in those repos, so we all avoid this going back and forth with our comments and waisting each other's time. Cheers!

**Comment by CaptainCodeman** [comment]
Well, you shouldn't go adding misinformation to people's issues uninvited and then ignoring the explanations of why it's unnecessary. Thankfully, you're not the GitHub police so I can chose to share or not share whatever I like. I took the time to create this repo to illustrate a couple of things for anyone who happens to find them useful. If you don't, or you can't understand it, just keep on walking.

Maybe if you take the time to re-read my comments and actually try the code suggested, instead of insisting that copy-paste code is the only way, you'll discover it works just fine.

**Comment by CaptainCodeman** [comment]
For anyone else reading, it boils down to the `.known` property indicating if the `.user` property can be relied on yet. So until the `.known` state is true, you shouldn't decide anything based on the state of `.user` as the SDK is still loading / checking the auth status.

Another way of doing it is to make the auth store contain a promise, so you would instead `await` the auth state being known and set.

The key part is it isn't instant and whatever method you're using, you have to wait for it to be set.

**Comment by kylesloper** [comment]
@CaptainCodeman do you have any suggestions how one would go about making a promise in the `auth` store. 
This is what I currently have in my `+page.server.js`:
```
import { redirect, error } from '@sveltejs/kit'

export const load = async ({ parent }) => {
  // session consists of just the user object, but could contain other preferences
  const session = await parent()
  if (!session.user) {
    throw redirect(307, '/sign-up')
  }
}
```

**Comment by CaptainCodeman** [comment]
See: https://www.captaincodeman.com/how-to-await-firebase-auth-with-sveltekit

But please don't hijack issues with non-related requests in future.

**Comment by kylesloper** [comment]
It is directly related actually since it queries your very unique and specific method in adopting Firebase. Looking through these issues it seems you really don't have the right attitude for opensource projects - I appreciate your generosity in making this repo public however like @idr4n suggested I think it would be best to turn issues off going forward. Kindness and tolerance cost nothing.
